fitbenchmarking/fitbenchmarking
# Add `minuit` and `dfo` to the regression tests

#### Description of Work
This PR adds the `minuit` software packages back into the regression tests. The `gradient_free` and `scipy_go` packages will not be added in this because they require bounds to be specified for their parameters.

The `dfo` and `minuit` packages have seen unreliable failures in the past, hence why they were removed, so I will try investigate why this is happening.

#### Investigation
After some investigation I found that running a minuit fit on only **one** of the unreliable problems would result in the same unreliable behaviour as previously, so the issue isn't caused by any 'interference' between instances of the `Minuit` python class. There were consistently two different answers appearing for the cost function of the `BENNETS5.SIF` dataset of `A = 2.113e-5` and `B = 2.477e-5`. However I could never replicate this unreliable issue whilst on the same machine during the same session (i.e. repeating the test multiple times did not result in different results either locally or remotely).

When you set the `print_level` to 2 or 3 you get more information about what is going on in the fit. It appears there is an `MnSeedGenerator` object created right at the beginning, and after this the calculation values for `A` and `B` begin to diverge slightly. By the end of the fitting process, the differences are large enough for the stopping condition to be triggered at a different iteration. The stopping condition is the `EDM` or `Estimated Distance to Minimum` which has a default value of 0.0002 for Least Squares (but migrad can violate this by a factor of 10):
https://github.com/scikit-hep/iminuit/blob/23be056e6d25b1c27df6f0529d5d050e5ed8fb2f/src/iminuit/minuit.py#L165

If you reduce the 'tolerance' value then you can decrease the `EDM` max and get more accurate results. Setting this to zero meant the minimum possible tolerance was used, and the results achieved for the `BENNETS5.SIF` dataset became reliable. However, the results for other datasets which had cost function values which were much smaller (~1e-16) were still unreliable.

The best theory I have for what is happening is that the `MnSeedGenerator` is causing very small differences in the calculations early on resulting in different cost function results by the end of the fit. After reading some of the Minuit documentation, it makes me think that we should be choosing our datasets carefully for Migrad fitting. The solution **might** be to have a category of regression test where the problems are simple (e.g. only one local minimum). Maybe the category could be called `Deterministic` or `Simple` or something similar.

#### Other things I tried
I ruled out a few things while investigating:
- [x] There was no interference between instances of the `Minuit` class (which I thought could be possible as it creates a python wrapper around a c++ `MnMigrad` object which I thought could potential hang around in memory and be reused by chance in a subsequent iteration).
- [x] Changing the strategy to 'Careful' didn't work well 
https://github.com/scikit-hep/iminuit/blob/23be056e6d25b1c27df6f0529d5d050e5ed8fb2f/src/iminuit/minuit.py#L180
- [x] I checked there wasn't a precision mis-match when doing the calculations 
https://github.com/scikit-hep/iminuit/blob/23be056e6d25b1c27df6f0529d5d050e5ed8fb2f/src/iminuit/minuit.py#L121
- [x] Tested for thread safety. The function provided to minuit needs to be thread safe when OpenMP is being used for calculations. We are using MPI so should be ok, but I checked anyway and I don't think this is the problem


Fixes #939

#### Testing Instructions

1. Make sure the regression tests pass and the results look sensible.

Function: Does the change do what it's supposed to?

Tests: Does it pass? Is there adequate coverage for new code?

Style: Is the coding style consistent? Is anything overly confusing?

Documentation: Is there a suitable change to documentation for this change?
